### PR TITLE
Connection management code

### PR DESCRIPTION
**Pull request by mmihic**
R: @xichen2020 @robskillington

### CODE REVIEW

File: connection_manager.go [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+	provider ClusterMappingProvider
+	closed   bool
+	wg       sync.WaitGroup
+}
+
+// GetConnection retrieves a connection to the given database and cluster
+func (cm *connectionManager) GetConnection(database, cluster string) (Connection, error) {
+	cm.log.Debugf("retrieving connection for %s:%s", database, cluster)
+	for {
+		if conn, err := cm.getExistingConn(database, cluster); conn != nil || err != nil {
```
Comment (robskillington) [inline connection_manager.go]: Should the second clause be `conn == nil || err != nil`?

File: connection_manager.go [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+
+	// If someone closed us while we were upgrading the lock, bail out
+	if cm.closed {
+		return nil, errClosed
+	}
+
+	// If someone else created it while we were upgrading the lock, return the
+	// new connection
+	if c := cm.conns[key]; c != nil {
+		cm.Unlock()
```
Comment (robskillington) [inline connection_manager.go:191]: Is it not worth attempting to get the existing connection using just a read lock first?  Or do we expected consumers to call this once per database & cluster pair and hold onto connections (as in very infrequently so not worth optimizing for)?

File: connection_manager.go [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+		conn.Close()
+		return errClosed
+	}
+
+	// Register the watch and connection
+	cm.conns[key], cm.watches[key] = conn, cw
+
+	// Spin up a goroutine to monitor the cluster for updates
+	go cm.watchCluster(cl, cw, conn)
+	cm.wg.Add(1)
```
Comment (robskillington) [inline connection_manager.go:281]: This `cm.wg.Add(1)` statement would technically be slightly more safe if performed before the `go cm.watchCluster(cl, cw, conn)` statement perhaps?

File: connection_manager.go [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+
+	return nil
+}
+
+// watchCluster monitors the given cluster for config updates
+func (cm *connectionManager) watchCluster(cl Cluster, cw ClusterWatch, conn Connection) {
+
```
Comment (robskillington) [inline connection_manager.go]: nit: whitespace endline at start of method.

File: connection_manager.go [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+		cm.reconfigureConn(cw.Get(), conn)
+	}
+
+	cm.cleanupConn(cl.Database(), cl.Name())
+	cm.wg.Done()
+}
+
+// reconfigureConn reconfigures a connection with new cluster configuration
+func (cm *connectionManager) reconfigureConn(cl Cluster, existing Connection) error {
+	key := fmtConnKey(cl.Database(), cl.Name())
```
Comment (robskillington) [inline connection_manager.go]: Would it be worth making a few different versions of `fmtConn...` with one top level version just taking a `Cluster` to make this slightly less manual work, i.e. `fmtConnKeyFromCluster(cl Cluster)` to save the double call each time?

File: connection_manager.go [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+	"github.com/m3db/m3x/close"
+	"github.com/m3db/m3x/log"
+
+	"github.com/facebookgo/clock"
+	"github.com/golang/protobuf/proto"
+)
+
+var (
+	errClosed          = errors.New("closed")
+	errTypeUnsupported = errors.New("unsupported storage type")
```
Comment (xichen2020) [inline connection_manager.go]: Perhaps `errStorageTypeUnsupported` is more clear?

File: connection_manager.go (hunk 2 above)
Comment (mmihic) [reply connection_manager.go:191]: We do - see the `GetConnection` code which is the main line for retrieving a connection .  You only enter `tryOpenConn` if there is no existing connection obtainable under a read lock, and you are going to try to open one.

File: connection_manager.go (hunk 3 above)
Comment (mmihic) [reply connection_manager.go:281]: Ah, yeah, bad refactoring.  Will fix.

File: connection_manager.go (hunk 6 above)
Comment (mmihic) [reply connection_manager.go]: Sure

File: connection_manager.go (hunk 5 above)
Comment (mmihic) [reply connection_manager.go]: Ok.

File: connection_manager.go [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+	if err := cm.registerConn(cl, cw, conn); err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// openConn opens a connection to the given cluster
+func (cm *connectionManager) openConn(cl Cluster) (Connection, error) {
+	d := cm.drivers[cl.Type().Name()]
```
Comment (xichen2020) [inline connection_manager.go:247]: Worth merging line 247-260 with line 301-317 into a separate method that gets reused in both `openConn` and `reconfigureConn`?

File: connection_manager.go [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+// cleanupConn cleans up the connection to the given database and cluster
+func (cm *connectionManager) cleanupConn(database, cluster string) {
+	cm.log.Infof("cleaning up connection to %s:%s", database, cluster)
+
+	key := fmtConnKey(database, cluster)
+
+	cm.Lock()
+	conn := cm.conns[key]
+	delete(cm.conns, key)
+	delete(cm.watches, key)
```
Comment (xichen2020) [inline connection_manager.go:340]: Do we need to close the watch here?

File: connection_manager.go (hunk 2 above)
Comment (robskillington) [reply connection_manager.go:191]: Ah right, this is lock promotion. :+1:

File: connection_manager.go (hunk 7 above)
Comment (mmihic) [reply connection_manager.go:247]: Sure

File: connection_manager.go [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+	}
+
+	newConn, err := d.ReconfigureConnection(existing, cfg)
+	if err != nil {
+		cm.log.Errorf("could not reconfigure conn for %s:%d v%d: %v",
+			cl.Database(), cl.Name(), cl.Config().Version(), err)
+		return err
+	}
+
+	cm.Lock()
```
Comment (xichen2020) [inline connection_manager.go:325]: Just checking, the read-write lock can be recursively locked from different go routines?

File: connection_manager.go (hunk 8 above)
Comment (mmihic) [reply connection_manager.go:340]: No, the watch is closed in `connectionManager.Close()`.  Closing the watch is what causes the background goroutine to break out of its watch loop and perform this cleanup code.

File: connection_manager.go (hunk 9 above)
Comment (mmihic) [reply connection_manager.go:325]: RWLocks cannot be recursively locked.  Not sure what you mean by "from different go routines" - multiple readers can acquire the read lock, but a write lock is exclusive to the writer.

Comment (xichen2020) [reply connection_manager.go:325]: Ah misread the code, my bad

File: connection_manager.go (hunk 8 above)
Comment (xichen2020) [reply connection_manager.go:340]: Ah I see, it's the other way around, sounds good 👍

**Pull request merged**